modules/extensions-support/src/main/java/org/mule/module/extension/internal/runtime/source/ExtensionMessageSource.java

```diff
@@ -102,10 +113,49 @@ public void handle(MuleMessage<Object, Serializable> message, CompletionHandler<
     }
 
     @Override
+    public void handle(MuleMessage<Object, Serializable> message)
+    {
+        MuleEvent event = new DefaultMuleEvent((org.mule.api.MuleMessage) message, REQUEST_RESPONSE, flowConstruct);
+        messageProcessingManager.processMessage(new ExtensionFlowProcessingTemplate(event, messageProcessor, new CompletionHandler<MuleMessage, MessagingException>()
+        {
+            @Override
+            public void onCompletion(MuleMessage result)
+            {
+                //Nothing to do
+            }
+
+            @Override
+            public void onFailure(MessagingException exception)
+            {
+                //Nothing to do
+            }
+        }), createProcessingContext());
+    }
+
+    //TODO: MULE-9397 - Add support for @OnException for sources
+    @Override
     public void onException(Throwable exception)
     {
-        LOGGER.warn(String.format("Message source '%s' on flow '%s' threw exception. Restarting...", source.getName(), flowConstruct.getName()), exception);
-        restartSource();
+        Optional<ConnectionException> connectionException = ExceptionUtils.extractRootConnectionException(exception);
+        try
+        {
+            if (connectionException.isPresent())
+            {
+                LOGGER.warn(String.format("Message source '%s' on flow '%s' threw exception. Restarting...", source.getName(), flowConstruct.getName()), exception);
+                stopSource();
+                disposeSource();
+                start();
+            }
+            else
+            {
+                throw exception;
+            }
+        }
+        catch (Throwable e)
```

Comment:
Yes, I'm doing that. I did it to have a unique point of catch, reusing the logging and shutdown logic